### PR TITLE
MessageKit removed, Drafty implemented.

### PR DESCRIPTION
**Pull request by or-else**

### CODE REVIEW

File: Tinodios/UiUtils.swift [hunk 1]
```diff
@@ -73,6 +72,22 @@ class UiUtils {
         // Reset red border to default.
         field.layer.borderWidth = 0.0
     }
+
+    public static func bytesToHumanSize(_ bytes: Int64) -> String {
+        if bytes <= 0 {
```
Comment (aforge) [inline Tinodios/UiUtils.swift]: guard bytes > 0 else { return "0 Bytes" }

File: TinodeSDK/Topic.swift [hunk 2]
```diff
@@ -778,30 +779,36 @@ open class Topic<DP: Codable, DR: Codable, SP: Codable, SR: Codable>: TopicProto
...
     }
     @discardableResult
     public func noteRecv() -> Int {
-        let result = noteReadRecv(what: NoteType.kRecv)
+        let result = note(what: NoteType.kRecv)
         store?.setRecv(topic: self, recv: result)
         return result
     }
+    public func noteKeyPress() {
+        _ = note(what: NoteType.kKeyPress)
```
[the remark this answers is not in the document]
Comment (or-else) [reply TinodeSDK/Topic.swift]: done

File: Tinodios/AttribFormatter.swift [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+        static let kMinButtonWidth: Int = 10
+        // Line hight multiplier in forms.
+        static let kFormLineSpacing: CGFloat = 1.5
+    }
+
+    typealias CharacterStyle = [NSAttributedString.Key: Any]
+
+    let defaultAttrs: [NSAttributedString.Key: Any]
+
+    init(defaultAttrs attrs: [NSAttributedString.Key: Any]) {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

File: Tinodios/AttribFormatter.swift [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+//
+//  Converts Drafty instance into attributed text suitable for display in UITextView
+
+// For MIME -> UTI conversion
+import MobileCoreServices
+import TinodeSDK
+import UIKit
+
+// iOS's support for styled strings is much weaker than Android's and web. Some styles cannot be nested. They have to be constructed and applied all at once at the leaf level.
+class AttribFormatter: DraftyFormatter {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: -> AttributedStringFormatter

File: Tinodios/AttribFormatter.swift [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    typealias CharacterStyle = [NSAttributedString.Key: Any]
+
+    let defaultAttrs: [NSAttributedString.Key: Any]
+
+    init(defaultAttrs attrs: [NSAttributedString.Key: Any]) {
+        defaultAttrs = attrs
+    }
+
+    // Inline image
+    private func handleImage(content: TreeNode, attr: [String : JSONValue]?) {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

File: Tinodios/AttribFormatter.swift [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+        attachment.button = url
+        content.attachment(attachment)
+    }
+
+    // Convert button payload to an URL.
+    // NSAttributedString.Key.link wants payload to be NSURL.
+    private static func buttonDataAsUri(face: TreeNode, attr: [String : JSONValue]?) -> String? {
+        guard let attr = attr, let actionType = attr["act"]?.asString() else { return nil }
+        var baseUrl: URLComponents
+        if actionType == "url" {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

File: Tinodios/AttribFormatter.swift [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+            return NSMutableAttributedString(string: content.string, attributes: attributes)
+        }
+
+        let formatTree = content.format(formatter: AttribFormatter(defaultAttrs: attributes))
+        return formatTree.toAttributed(defaultAttrs: attributes, fontTraits: nil, size: maxSize)
+    }
+
+    // File or image attachment.
+    struct Attachment {
+        var data: Data?
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

File: Tinodios/AttribFormatter.swift [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+        }
+
+        func addNode(content: String?) {
+            guard let content = content else { return }
+
+            addNode(node: TreeNode(content: content))
+        }
+
+        var isEmpty: Bool {
+            return (text == nil || text!.isEmpty) && (children == nil || children!.isEmpty)
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

File: Tinodios/MessageBubbleDecorator.swift [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    private static func drawIncomig(path: UIBezierPath, size: CGSize, style: Style) {
+        let width = size.width
+        let height = size.height
+
+        // Start at bottom-left corner.
+        if style == .single || style == .last {
+            // Leave space for the rounded corner.
+            path.move(to: CGPoint(x: 22, y: height))
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageBubbleDecorator.swift]: Visually. These values make a nice message bubble. It seems to be a common practice. For example, MessageKit uses constants like this for corner radius. I don't think there is much value in parametrising the bubble further.

File: Tinodios/MessageCell.swift [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        timestampLabel.text = nil
+        deliveryMarker.image = nil
+        avatarView.image = nil
+    }
+
+    /// Handle tap gesture on contentView and its subviews.
+    func handleTapGesture(_ gesture: UIGestureRecognizer) {
+        let touchLocation = gesture.location(in: self)
+
+        switch true {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageCell.swift]: :)

File: Tinodios/ChatListViewController.swift [hunk 11]
```diff
@@ -117,3 +118,24 @@ extension ChatListViewController {
         return cell
     }
 }
+
+extension ChatListViewController {
+
+    /// Show notification that the chat list is empty
+    public func showNoChatsNote(_ show: Bool) {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/ChatListViewController.swift]: done

File: Tinodios/MessageViewController.swift [hunk 12]
```diff
@@ -184,171 +289,433 @@ extension MessageViewController {
...
-extension MessageViewController: MessagesDataSource {
-    func currentSender() -> Sender {
-        return Sender(id: myUID ?? "???", displayName: "")
-    }
-    func numberOfSections(in messagesCollectionView: MessagesCollectionView) -> Int {
+// Methods for filling message with content and layout out message subviews.
+extension MessageViewController: UICollectionViewDataSource {
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        self.collectionView.showNoMessagesNote(messages.count == 0)
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageViewController.swift]: done + showNoMessagesNote -> toggleNoMessagesNote(on:)

File: Tinodios/UiUtils.swift (hunk 1 above)
Comment (or-else) [reply Tinodios/UiUtils.swift]: done

File: Tinodios/AttribFormatter.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: and also renamed in a bunch of other places

File: Tinodios/AttribFormatter.swift [hunk 13]
```diff
@@ -0,0 +1,500 @@
...
+                attributed.addAttributes([.paragraphStyle: pstyle], range: NSRange(location: 0, length: attributed.length))
+            }
+
+            attributed.endEditing()
+            return attributed
+        }
+    }
+}
+
+extension StoredMessage {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/AttribFormatter.swift]: done

**Pull request merged**